engine/src/xcc.rs

```diff
@@ -0,0 +1,172 @@
+                address: sender,
+                version: latest_code_version,
+            };
+            let callback = PromiseCreateArgs {
+                target_account_id: current_account_id.clone(),
+                method: "factory_update_address_version".into(),
+                args: args.try_to_vec().unwrap(),
+                attached_balance: ZERO_YOCTO,
+                attached_gas: VERSION_UPDATE_GAS,
+            };
+
+            handler.promise_attach_callback(promise_id, &callback)
+        }
+        AddressVersionStatus::UpToDate => handler.promise_create_call(&promise),
+    };
+}
+
+/// Read the current wasm bytecode for the router contracts
+pub fn get_router_code<I: IO>(io: &I) -> RouterCode {
+    let key = storage::bytes_to_key(KeyPrefix::CrossContractCall, CODE_KEY);
+    let bytes = io.read_storage(&key).expect(ERR_NO_ROUTER_CODE).to_vec();
+    RouterCode(bytes)
+}
+
+/// Set new router bytecode, and update increment the version by 1.
+pub fn update_router_code<I: IO>(io: &mut I, code: &RouterCode) {
+    let key = storage::bytes_to_key(KeyPrefix::CrossContractCall, CODE_KEY);
+    io.write_storage(&key, &code.0);
+
+    let current_version = get_latest_code_version(io);
+    set_latest_code_version(io, current_version.increment());
+}
+
+/// Get the latest router contract version.
+pub fn get_latest_code_version<I: IO>(io: &I) -> CodeVersion {
+    let key = storage::bytes_to_key(KeyPrefix::CrossContractCall, VERSION_KEY);
+    read_version(io, &key).unwrap_or_default()
+}
+
+/// Get the version of the currently deploy router for the given address (if it exists).
+pub fn get_code_version_of_address<I: IO>(io: &I, address: &Address) -> Option<CodeVersion> {
+    let key = storage::bytes_to_key(KeyPrefix::CrossContractCall, address.as_bytes());
+    read_version(io, &key)
+}
+
+/// Set the version of the router contract deployed for the given address.
+pub fn set_code_version_of_address<I: IO>(io: &mut I, address: &Address, version: CodeVersion) {
+    let key = storage::bytes_to_key(KeyPrefix::CrossContractCall, address.as_bytes());
+    let value_bytes = version.0.to_le_bytes();
+    io.write_storage(&key, &value_bytes);
+}
+
+/// Sets the latest router contract version. This function is intentionally private because
+/// it should never be set manually. The version is managed automatically by `update_router_code`.
+fn set_latest_code_version<I: IO>(io: &mut I, version: CodeVersion) {
+    let key = storage::bytes_to_key(KeyPrefix::CrossContractCall, VERSION_KEY);
+    let value_bytes = version.0.to_le_bytes();
+    io.write_storage(&key, &value_bytes);
+}
+
+/// Private utility method for reading code version from storage.
+fn read_version<I: IO>(io: &I, key: &[u8]) -> Option<CodeVersion> {
+    match io.read_u32(key) {
+        Ok(value) => Some(CodeVersion(value)),
+        Err(ReadU32Error::MissingValue) => None,
+        Err(ReadU32Error::InvalidU32) => panic!("{}", ERR_CORRUPTED_STORAGE),
```

Comment:
Yes, I was allowing panics in places where the only way the code reaches that point is due to a critical bug (so we want to fail fast to avoid more incorrect operations). The reason I am not using NEAR panics is because the standalone engine should also panic at this spot, not just the contract.